### ISSUE TITLE
TypeError in scv.pp.filter_and_normalize

### ISSUE THREAD

**Issue by plath997** [theislab/scvelo#290]
Hello,

I have a very large dataset (~400k cells) that I have been trying to run this implementation on. However, every time that I run scv.pp.filter_and_normalize using n_top_genes, I always get this error:

```TypeError: Cannot cast array data from dtype('float64') to dtype('<U32') according to the rule 'safe' ```

Would appreciate any work around for this issue. Thanks!

**Comment by VolkerBergen** [comment]
Please send the full error traceback, so to understand at which code line it breaks.

**Comment by plath997** [comment]
```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/.local/lib/python3.6/site-packages/scvelo/preprocessing/utils.py", line 821, in filter_and_normalize
    adata, n_top_genes=n_top_genes, retain_genes=retain_genes, flavor=flavor
  File "/home/.local/lib/python3.6/site-packages/scvelo/preprocessing/utils.py", line 488, in filter_genes_dispersion
    disp_std_bin[one_gene_per_bin] = disp_mean_bin[one_gene_per_bin].values
  File "/home/.local/lib/python3.6/site-packages/pandas/core/series.py", line 905, in __getitem__
    if is_iterator(key):
  File "/home/.local/lib/python3.6/site-packages/pandas/core/dtypes/inference.py", line 157, in is_iterator
    return hasattr(obj, '__next__')
  File "/home/.local/lib/python3.6/site-packages/pandas/core/generic.py", line 5065, in __getattr__
    if self._info_axis._can_hold_identifiers_and_holds_name(name):
  File "/home/.local/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 3985, in _can_hold_identifiers_and_holds_name
    return name in self
  File "/home/.local/lib/python3.6/site-packages/pandas/core/indexes/category.py", line 327, in __contains__
    return contains(self, key, container=self._engine)
  File "/home/.local/lib/python3.6/site-packages/pandas/core/arrays/categorical.py", line 188, in contains
    loc = cat.categories.get_loc(key)
  File "/home/.local/lib/python3.6/site-packages/pandas/core/indexes/interval.py", line 770, in get_loc
    start, stop = self._find_non_overlapping_monotonic_bounds(key)
  File "/home/.local/lib/python3.6/site-packages/pandas/core/indexes/interval.py", line 717, in _find_non_overlapping_monotonic_bounds
    start = self._searchsorted_monotonic(key, 'left')
  File "/home/.local/lib/python3.6/site-packages/pandas/core/indexes/interval.py", line 681, in _searchsorted_monotonic
    return sub_idx._searchsorted_monotonic(label, side)
  File "/home/.local/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 4756, in _searchsorted_monotonic
    return self.searchsorted(label, side=side)
  File "/home/.local/lib/python3.6/site-packages/pandas/core/base.py", line 1501, in searchsorted
    return self._values.searchsorted(value, side=side, sorter=sorter)
TypeError: Cannot cast array data from dtype('float64') to dtype('<U32') according to the rule 'safe'
```

**Comment by kathleenabadie** [comment]
Hello, I am having the same error and would be grateful for advice. My error traceback is as follows:

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-76-ec36098be2aa> in <module>
      1 
----> 2 scv.pp.filter_and_normalize(adata, min_counts=20, min_counts_u=10, n_top_genes=3000 )
      3 # getting error when I include n_top_genes parameter

~/anaconda2/lib/python3.6/site-packages/scvelo/preprocessing/utils.py in filter_and_normalize(data, min_counts, min_counts_u, min_cells, min_cells_u, min_shared_counts, min_shared_cells, n_top_genes, retain_genes, flavor, log, layers_normalize, copy, **kwargs)
    819     if n_top_genes is not None:
    820         filter_genes_dispersion(
--> 821             adata, n_top_genes=n_top_genes, retain_genes=retain_genes, flavor=flavor
    822         )
    823 

~/anaconda2/lib/python3.6/site-packages/scvelo/preprocessing/utils.py in filter_genes_dispersion(data, flavor, min_disp, max_disp, min_mean, max_mean, n_bins, n_top_genes, retain_genes, log, subset, copy)
    486                 gen_indices = np.where(one_gene_per_bin[df["mean_bin"].values])
    487                 gen_indices = gen_indices[0].tolist()
--> 488                 disp_std_bin[one_gene_per_bin] = disp_mean_bin[one_gene_per_bin].values
    489                 disp_mean_bin[one_gene_per_bin] = 0
    490                 # normalized dispersion

~/anaconda2/lib/python3.6/site-packages/pandas/core/series.py in __getitem__(self, key)
    903             raise
    904 
--> 905         if is_iterator(key):
    906             key = list(key)
    907 

~/anaconda2/lib/python3.6/site-packages/pandas/core/dtypes/inference.py in is_iterator(obj)
    155         # Python 3 generators have
    156         # __next__ instead of next
--> 157         return hasattr(obj, '__next__')
    158 
    159 

~/anaconda2/lib/python3.6/site-packages/pandas/core/generic.py in __getattr__(self, name)
   5063             return object.__getattribute__(self, name)
   5064         else:
-> 5065             if self._info_axis._can_hold_identifiers_and_holds_name(name):
   5066                 return self[name]
   5067             return object.__getattribute__(self, name)

~/anaconda2/lib/python3.6/site-packages/pandas/core/indexes/base.py in _can_hold_identifiers_and_holds_name(self, name)
   3983         """
   3984         if self.is_object() or self.is_categorical():
-> 3985             return name in self
   3986         return False
   3987 

~/anaconda2/lib/python3.6/site-packages/pandas/core/indexes/category.py in __contains__(self, key)
    325             return self.hasnans
    326 
--> 327         return contains(self, key, container=self._engine)
    328 
    329     @Appender(_index_shared_docs['contains'] % _index_doc_kwargs)

~/anaconda2/lib/python3.6/site-packages/pandas/core/arrays/categorical.py in contains(cat, key, container)
    186     #  can't be in container either.
    187     try:
--> 188         loc = cat.categories.get_loc(key)
    189     except KeyError:
    190         return False

~/anaconda2/lib/python3.6/site-packages/pandas/core/indexes/interval.py in get_loc(self, key, method)
    768                 key = self._maybe_cast_slice_bound(key, 'left', None)
    769 
--> 770             start, stop = self._find_non_overlapping_monotonic_bounds(key)
    771 
    772             if start is None or stop is None:

~/anaconda2/lib/python3.6/site-packages/pandas/core/indexes/interval.py in _find_non_overlapping_monotonic_bounds(self, key)
    715             # scalar or index-like
    716 
--> 717             start = self._searchsorted_monotonic(key, 'left')
    718             stop = self._searchsorted_monotonic(key, 'right')
    719         return start, stop

~/anaconda2/lib/python3.6/site-packages/pandas/core/indexes/interval.py in _searchsorted_monotonic(self, label, side, exclude_label)
    679                 label = _get_prev_label(label)
    680 
--> 681         return sub_idx._searchsorted_monotonic(label, side)
    682 
    683     def _get_loc_only_exact_matches(self, key):

~/anaconda2/lib/python3.6/site-packages/pandas/core/indexes/base.py in _searchsorted_monotonic(self, label, side)
   4754     def _searchsorted_monotonic(self, label, side='left'):
   4755         if self.is_monotonic_increasing:
-> 4756             return self.searchsorted(label, side=side)
   4757         elif self.is_monotonic_decreasing:
   4758             # np.searchsorted expects ascending sort order, have to reverse

~/anaconda2/lib/python3.6/site-packages/pandas/core/base.py in searchsorted(self, value, side, sorter)
   1499     def searchsorted(self, value, side='left', sorter=None):
   1500         # needs coercion on the key (DatetimeIndex does already)
-> 1501         return self._values.searchsorted(value, side=side, sorter=sorter)
   1502 
   1503     def drop_duplicates(self, keep='first', inplace=False):

TypeError: Cannot cast array data from dtype('float64') to dtype('<U32') according to the rule 'safe'

**Comment by VolkerBergen** [comment]
Trying again now, after installing from source `pip install git+https://github.com/theislab/scvelo`.

**Comment by kathleenabadie** [comment]
Thank you, that fixed it! I appreciate the quick help.
Best,
Kathleen

**Comment by WeilerP** [comment]
@plath997, did [this](https://github.com/theislab/scvelo/commit/8351f46922903f20eee61def96e7cc7af47a0475) resolve your issue? The fix is now included in `scvelo==0.2.3`. I'll close the issue for now. Happy to reopen if the issue still exists.
